### PR TITLE
[HttpClient] Support for cURL handler objects

### LINKED ISSUE #37378: [HttpClient] cURL handles are not resources in php8
**Symfony version(s) affected**: 5.1.2

**Description**  
A console tool that I'm building uses the HttpClient component to query an API. It works perfectly fine of php 7.4, but if I run it on php 8 with `ext-curl`, I get the following error:

```
TypeError {#81
  #message: "curl_setopt(): Argument #1 ($handle) must be of type CurlHandle, null given"
  #code: 0
  #file: "./vendor/symfony/http-client/Response/CurlResponse.php"
  #line: 72
  trace: {
    ./vendor/symfony/http-client/Response/CurlResponse.php:72 { …}
    ./vendor/symfony/http-client/CurlHttpClient.php:311 { …}
    ./path/to/my/code { …}
  }
}
```

**How to reproduce**  
The problem is apparently, that the code assumes that "curl handles" returned by `curl_init()` and `curl_multi_init()` are resources. However, in php 8 they're objects of type `CurlHandle`.

**Possible Solution**  
Adjust `CurlHttpClient` and related classes to the new type of curl handles. I could work on a patch.

**Additional Context**
`is_resource()` is called on a handle:
https://github.com/symfony/symfony/blob/2c9e9cd0c20f5206003948c63ba027ae760e060b/src/Symfony/Component/HttpClient/Response/CurlResponse.php#L46

A handle is cast to an int which does not work on objects:
https://github.com/symfony/symfony/blob/2c9e9cd0c20f5206003948c63ba027ae760e060b/src/Symfony/Component/HttpClient/Response/CurlResponse.php#L60

### PR DESCRIPTION
**Pull request by derrabus**
| Q             | A
| ------------- | ---
| Branch?       | 4.4
| Bug fix?      | yes
| New feature?  | no
| Deprecations? | no
| Tickets       | Fix #37378
| License       | MIT
| Doc PR        | N/A

### CODE REVIEW

**Comment by derrabus** [comment]
Note: The php 8 build on Travis will fail until https://github.com/php/php-src/commit/26171c3690874743e1c71766e60248fbd974782f lands in their nightly build. I've run the tests locally.

File: src/Symfony/Component/HttpClient/CurlHttpClient.php [hunk 1]
```diff
@@ -347,7 +347,7 @@ public function reset()
         }
 
         foreach ($this->multi->openHandles as [$ch]) {
-            if (\is_resource($ch)) {
+            if (\is_resource($ch) || $ch instanceof \CurlHandle) {
```
Comment (nikic) [inline src/Symfony/Component/HttpClient/CurlHttpClient.php:350]: Just wondering, why are these is_resource checks necessary in the first place? Are they getting closed somehow without being removed?

Comment (derrabus) [reply src/Symfony/Component/HttpClient/CurlHttpClient.php:350]: Good question indeed. I haven't found a reason why `$ch` should be anything else than a cURL handle. And the tests still pass if I remove the two guard clauses.

Maybe @nicolas-grekas knows more about why we need(ed?) this check.

Comment (nicolas-grekas) [reply src/Symfony/Component/HttpClient/CurlHttpClient.php:350]: I've seen strange things happen at destructing/shutdown time, this might be related.

**Review by nicolas-grekas — APPROVED** [review]

**Review by fabpot — APPROVED** [review]

**Comment by fabpot** [comment]
Thank you @derrabus.

**Pull request merged**